### ISSUE TITLE
Gradle hides some details of jvm crashes

### ISSUE THREAD

**Issue by rjernst** [elastic/elasticsearch#14847]
If the jvm crashes during a test run, we get limited output like this:

```
[ant:junit4] ERROR: JVM J3 ended with an exception, command line: /usr/local/jenkins/tools/hudson.model.JDK/JDK8/bin/java -ea -Xmx1024m -Xms1024m -XX:MaxDirectMemorySize=512m -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/mnt/jenkins/workspace/es_core_master_strong/core/build/heapdump -server -XX:+UseG1GC -XX:-UseCompressedOops -XX:+AggressiveOpts -Dtests.prefix=tests -Dtests.seed=AECA7AEAD5CCBA53 -Dtests.security.manager=true -Djava.io.tmpdir=./temp -Dtests.maven=true -Dtests.artifact=core -Dtests.heap.size=1024m "-Dtests.jvm.argline=-server -XX:+UseG1GC -XX:-UseCompressedOops -XX:+AggressiveOpts" -Djava.awt.headless=true -Dtests.ifNoTests=fail -Dtests.nightly=false -Des.logger.level=DEBUG -Dtests.task=:core:integTest -Dtests.network=false -Djunit4.childvm.cwd=/mnt/jenkins/workspace/es_core_master_strong/core/build/testrun/integTest/J3 -Djunit4.childvm.id=3 -Djunit4.childvm.count=4 -classpath /mnt/jenkins/workspace/es_core_master_strong/core/build/classes/test:/mnt/jenkins/workspace/es_core_master_strong/core/build/resources/test:/mnt/jenkins/workspace/es_core_master_strong/core/build/classes/main:/mnt/jenkins/workspace/es_core_master_strong/core/build/resources/main:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-core/5.4.0-snapshot-1714615/d8d3f55ea169ec7a483c076d8d786d2fa7ec0447/lucene-core-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-analyzers-common/5.4.0-snapshot-1714615/95e58f09a8878e74f2cceba8aca898d7464782a8/lucene-analyzers-common-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-backward-codecs/5.4.0-snapshot-1714615/bf7600f7556e4cbdfb7b93947e0ef2312d4bfa6f/lucene-backward-codecs-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-grouping/5.4.0-snapshot-1714615/a77cedd6e706317f2f3c1b693736ed191a30a488/lucene-grouping-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-highlighter/5.4.0-snapshot-1714615/84d9d5617fae821a7cf1f0f5ac961f4128a936e9/lucene-highlighter-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-join/5.4.0-snapshot-1714615/9e7646fb8cfbb72e061eba83467a05671cfa5ee1/lucene-join-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-memory/5.4.0-snapshot-1714615/68417b06ee7adf7f751bbeabdb2ed14e5ef3172/lucene-memory-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-misc/5.4.0-snapshot-1714615/f12de7f380e1f40b9432c90a2bb0b9ba67e94c41/lucene-misc-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-queries/5.4.0-snapshot-1714615/8313aa54e3d13e40bf7bf32b6ee68c7c4b547d6a/lucene-queries-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-queryparser/5.4.0-snapshot-1714615/62696f81612a2594c5eac93c9db57c108d3e951a/lucene-queryparser-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-sandbox/5.4.0-snapshot-1714615/a506a09e97070b74c226627ab8adcde9f32576ef/lucene-sandbox-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-spatial/5.4.0-snapshot-1714615/22924c220508d02a30e9425f6293fcff1621a23a/lucene-spatial-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-spatial3d/5.4.0-snapshot-1714615/af3be612c990303f6c575411b436c967f665876c/lucene-spatial3d-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-suggest/5.4.0-snapshot-1714615/c797cf414ac18868e804afe6be65c30046a5d2dc/lucene-suggest-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.elasticsearch/securesm/1.0/c0c6cf986ba0057390bfcc80c366a0e3157f944b/securesm-1.0.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/commons-cli/commons-cli/1.3.1/1303efbc4b181e5a58bf2e967dc156a3132b97c0/commons-cli-1.3.1.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.carrotsearch/hppc/0.7.1/8b5057f74ea378c0150a1860874a3ebdcb713767/hppc-0.7.1.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/joda-time/joda-time/2.8.2/d27c24204c5e507b16fec01006b3d0f1ec42aed4/joda-time-2.8.2.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.joda/joda-convert/1.2/35ec554f0cd00c956cc69051514d9488b1374dec/joda-convert-1.2.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.core/jackson-core/2.6.2/123f29333b2c6b3516b14252b6e93226bfcd6e37/jackson-core-2.6.2.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.dataformat/jackson-dataformat-smile/2.6.2/395d18c1a1dd730b8026ee59c4067e5d2b45ba6e/jackson-dataformat-smile-2.6.2.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.dataformat/jackson-dataformat-yaml/2.6.2/4ae23088dd3fae47c66843f2e4251d7255ee140e/jackson-dataformat-yaml-2.6.2.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.dataformat/jackson-dataformat-cbor/2.6.2/1e13c575f914c83761bb8e2aca7dfd9e4c647579/jackson-dataformat-cbor-2.6.2.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.yaml/snakeyaml/1.15/3b132bea69e8ee099f416044970997bde80f4ea6/snakeyaml-1.15.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/io.netty/netty/3.10.5.Final/9ca7d55d246092bddd29b867706e2f6c7db701a0/netty-3.10.5.Final.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.ning/compress-lzf/1.0.2/62896e6fca184c79cc01a14d143f3ae2b4f4b4ae/compress-lzf-1.0.2.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.tdunning/t-digest/3.0/84ccf145ac2215e6bfa63baa3101c0af41017cfc/t-digest-3.0.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.hdrhistogram/HdrHistogram/2.1.6/7495feb7f71ee124bd2a7e7d83590e296d71d80e/HdrHistogram-2.1.6.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.spatial4j/spatial4j/0.5/6e16edaf6b1ba76db7f08c2f3723fce3b358ecc3/spatial4j-0.5.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.vividsolutions/jts/1.13/3ccfb9b60f04d71add996a666ceb8902904fd805/jts-1.13.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.github.spullara.mustache.java/compiler/0.9.1/14aec5344639782ee76441401b773946c65eb2b3/compiler-0.9.1.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/log4j/log4j/1.2.17/5af35056b4d257e4b64b9e8069c0746e8b08629f/log4j-1.2.17.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/log4j/apache-log4j-extras/1.2.17/85863614d82185d7e51fe21c00aa9117a523a8b6/apache-log4j-extras-1.2.17.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-api/1.6.2/8619e95939167fb37245b5670135e4feb0ec7d50/slf4j-api-1.6.2.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/net.java.dev.jna/jna/4.1.0/1c12d070e602efd8021891cdd7fd18bc129372d4/jna-4.1.0.jar:/mnt/jenkins/workspace/es_core_master_strong/test-framework/build/libs/test-framework-3.0.0-SNAPSHOT.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.carrotsearch.randomizedtesting/randomizedtesting-runner/2.2.0/60de504132241be049564a3a34fd7dcc296e2ef0/randomizedtesting-runner-2.2.0.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/junit/junit/4.11/4e031bb61df09069aeb2bffb4019e7a5034a4ee0/junit-4.11.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.hamcrest/hamcrest-all/1.3/63a21ebc981131004ad02e0434e799fd7f3a8d5a/hamcrest-all-1.3.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-test-framework/5.4.0-snapshot-1714615/1e4bf816dd65707f0b2d8041af63e91df71cb7a0/lucene-test-framework-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.lucene/lucene-codecs/5.4.0-snapshot-1714615/3e324496abc07c4d32f2492a7941dd01824af734/lucene-codecs-5.4.0-snapshot-1714615.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.httpcomponents/httpclient/4.3.6/4c47155e3e6c9a41a28db36680b828ced53b8af4/httpclient-4.3.6.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/org.apache.httpcomponents/httpcore/4.3.3/f91b7a4aadc5cf486df6e4634748d7dd7a73f06d/httpcore-4.3.3.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/commons-logging/commons-logging/1.1.3/f6f66e966c70a83ffbdb6f17a0919eaf7c8aca7f/commons-logging-1.1.3.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/commons-codec/commons-codec/1.10/4b95f4897fa13f2cd904aee711aeafc0c5295cd8/commons-codec-1.10.jar:/var/lib/jenkins/.gradle/caches/modules-2/files-2.1/com.carrotsearch.randomizedtesting/junit4-ant/2.2.0/d401c9c729deccd5db8a5df3102eb18793c2224/junit4-ant-2.2.0.jar com.carrotsearch.ant.tasks.junit4.slave.SlaveMainSafe -eventsfile /mnt/jenkins/workspace/es_core_master_strong/core/build/testrun/integTest/temp/junit4-J3-20151118_090500_411.events @/mnt/jenkins/workspace/es_core_master_strong/core/build/testrun/integTest/temp/junit4-J3-20151118_090500_411.suites -stdin
[ant:junit4] ERROR: JVM J3 ended with an exception: Forked process returned with error code: 137.
	at com.carrotsearch.ant.tasks.junit4.JUnit4.executeSlave(JUnit4.java:1483)
	at com.carrotsearch.ant.tasks.junit4.JUnit4.access$000(JUnit4.java:130)
	at com.carrotsearch.ant.tasks.junit4.JUnit4$2.call(JUnit4.java:963)
	at com.carrotsearch.ant.tasks.junit4.JUnit4$2.call(JUnit4.java:960)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

But we dont see the stderr from the jvm that crashed. This is either something we are omitting in the gradle logger that connects to junit4's events, or we are misconfiguring junit4.

**Comment by rjernst** [comment]
I've opened #14850 to improve various aspects of how output is handled for tests. I don't think it handles issue though, as with --info logging (which we have in jenkins) we are already getting the INFO ant messages, which is what extraneous jvm output is logged at by junit4. I did improve the case when *not* running with --info, so that you will still see this output.

I also investigated piping stderr and stdout directly. This doesn't work for extraneous jvm output, since we have no control over that (things the jvm outputs directly, disregarding when System.out/err is changed). This is what junit4 logs on its own, and I've opened an issue there to improve the situation: randomizedtesting/randomizedtesting#222

For regular test stderr/stdout, the problem with not using gradle's logging is the stderr and stdout is captured while ant is running. Stderr actually comes through (as ant log messages) but stdout does not, when we are running with gradle's normal logging level (both appear when run with --info). I did fix a number of things in the referenced PR, though, which I think will improve the situation greatly. But unless we can actually reproduce the crash (or even know why it crashed and thus what was expected to be output that would indicate the reason for the crash), I don't know what else we could do here.

For reference, this is the original jenkins job that failed:
http://build-us-00.elastic.co/job/es_core_master_strong/5550/

FWIW, I looked through maven jenkins jobs for crashes, and found this one, which has no more information than that which we saw with gradle:
http://build-us-00.elastic.co/job/es_core_2x_metal/454/console
